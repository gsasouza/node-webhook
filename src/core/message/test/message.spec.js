const {
  getQueues,
  clearQueues,
} = require('../../../test/mock');

const testSetup = require('../../../test/setup');

const {
  addNewMessageToQueue,
  consumeMessages,
  consumeRetryMessages,
} = require('../message');

const { publish } = require('../../amqp');

beforeAll(() => testSetup());

beforeEach(() => clearQueues());

test('Should add a new message to queue', async () => {
  await addNewMessageToQueue({
    webhook: 'https://localhost.com:5003',
    payload: {
      message: 'test',
    }
  });
  expect(getQueues()).toMatchSnapshot();
});

test('Should consume and delivery a first try message', async () => {
  await addNewMessageToQueue({
    webhook: 'successUrl',
    payload: {
      message: 'test',
    }
  });
  await consumeMessages();
  expect(getQueues()).toMatchSnapshot();
});

test('Should consume and get an error on delivery with a first try message', async () => {
  await addNewMessageToQueue({
    webhook: 'errorUrl',
    payload: {
      message: 'test',
    }
  });
  await consumeMessages();
  expect(getQueues()).toMatchSnapshot();
});

test('Should get a retry message that is in idle time', async () => {
  const updatedAt = new Date(2018, 8, 2, 1, 0, 0, 0);
  Date.now = jest.fn(() => new Date(2018, 8, 2, 1, 14, 1, 0));
  await publish(3,
    {
      payload: {
        message: 'test',
      },
      webhook: 'https://localhost.com:5003',
      tryTimes: 0,
      updatedAt,
    }
  );
  await consumeRetryMessages();
  expect(getQueues()).toMatchSnapshot();
});

test('Should consume and delivery with a retry message', async () => {
  const updatedAt = new Date(2018, 8, 2, 1, 0, 0, 0);
  Date.now = jest.fn(() => new Date(2018, 8, 2, 1, 15, 1, 0));
  await publish(3,
    {
      payload: {
        message: 'test',
      },
      webhook: 'successUrl',
      tryTimes: 2,
      updatedAt,
    }
  );
  await consumeRetryMessages();
  expect(getQueues()).toMatchSnapshot();
});

test('Should consume and get an error on delivery with a retry message', async () => {
  const updatedAt = new Date(2018, 8, 2, 1, 0, 0, 0);
  Date.now = jest.fn(() => new Date(2018, 8, 2, 1, 15, 1, 0));
  await publish(3,
    {
      payload: {
        message: 'test',
      },
      webhook: 'errorUrl',
      tryTimes: 2,
      updatedAt,
    }
  );
  await consumeRetryMessages();
  expect(getQueues()).toMatchSnapshot();
});
