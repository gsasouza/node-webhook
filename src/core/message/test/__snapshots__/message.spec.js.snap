// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should add a new message to queue 1`] = `
Object {
  "1": Array [
    "{\\"webhook\\":\\"https://localhost.com:5003\\",\\"payload\\":{\\"message\\":\\"test\\"},\\"tryTimes\\":0,\\"messageId\\":1,\\"updatedAt\\":1}",
  ],
  "2": Array [],
  "3": Array [],
  "4": Array [],
  "5": Array [],
  "6": Array [],
  "7": Array [],
}
`;

exports[`Should consume and delivery a first try message 1`] = `
Object {
  "1": Array [],
  "2": Array [],
  "3": Array [],
  "4": Array [],
  "5": Array [],
  "6": Array [],
  "7": Array [],
}
`;

exports[`Should consume and delivery with a retry message 1`] = `
Object {
  "1": Array [],
  "2": Array [],
  "3": Array [],
  "4": Array [],
  "5": Array [],
  "6": Array [],
  "7": Array [],
}
`;

exports[`Should consume and get an error on delivery with a first try message 1`] = `
Object {
  "1": Array [],
  "2": Array [
    "{\\"webhook\\":\\"errorUrl\\",\\"payload\\":{\\"message\\":\\"test\\"},\\"tryTimes\\":1,\\"messageId\\":1,\\"updatedAt\\":1}",
  ],
  "3": Array [],
  "4": Array [],
  "5": Array [],
  "6": Array [],
  "7": Array [],
}
`;

exports[`Should consume and get an error on delivery with a retry message 1`] = `
Object {
  "1": Array [],
  "2": Array [],
  "3": Array [],
  "4": Array [
    "{\\"payload\\":{\\"message\\":\\"test\\"},\\"webhook\\":\\"errorUrl\\",\\"tryTimes\\":3,\\"updatedAt\\":\\"2018-09-02T04:15:01.000Z\\"}",
  ],
  "5": Array [],
  "6": Array [],
  "7": Array [],
}
`;

exports[`Should get a retry message that is in idle time 1`] = `
Object {
  "1": Array [],
  "2": Array [
    "{\\"payload\\":{\\"message\\":\\"test\\"},\\"webhook\\":\\"https://localhost.com:5003\\",\\"tryTimes\\":1,\\"updatedAt\\":\\"2018-09-02T04:14:01.000Z\\"}",
  ],
  "3": Array [],
  "4": Array [],
  "5": Array [],
  "6": Array [],
  "7": Array [],
}
`;
